### PR TITLE
[GOBBLIN-1941] Develop Temporal abstractions, including `Workload` for workflows of unbounded size through sub-workflow nesting

### PR DESCRIPTION
**Pull request by phet**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [ ] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1941

    "Create abstractions for developing on Temporal"

### Description
- [ ] Here are some details about my PR, including screenshots (if applicable):

Refine Temporal abstractions, including:
* `Workload` for workflows of unbounded size through sub-workflow nesting
* `WorkerConfig`, for tunneling the `TemporalWorker`'s configuration to the workflows and activities it hosts

### Tests
- [ ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

manual execution

### Commits
- [ ] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/SeqBackedWorkSpan.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    if (statefulDelegatee == null) {
+      statefulDelegatee = elems.iterator();
+    }
+    return statefulDelegatee.hasNext();
+  }
+
+  @Override
+  public WORK_ITEM next() {
+    if (statefulDelegatee == null) {
+      throw new IllegalStateException("first call `hasNext()`!");
```
Comment (Will-Lo) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/SeqBackedWorkSpan.java:56]: Is it regular for collections to enforce that `next()` is only called after `hasNext()`? When I was using this class prior I was tripped up around here as my expectation was that next() will fail only if there are no elements left in the list

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/WFAddr.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+
+/** Hierarchical address for nesting workflows (0-based). */
+@NoArgsConstructor // IMPORTANT: for jackson (de)serialization
+@RequiredArgsConstructor
+public class WFAddr {
```
Comment (Will-Lo) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/WFAddr.java:34]: Is it okay if we name if `WorkFlowAddr`? Shortening it seems unnecessary here given we refer to work and workflows in other classes without shortening them

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/workflow/AbstractNestingExecWorkflowImpl.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import io.temporal.workflow.Async;
+import io.temporal.workflow.ChildWorkflowOptions;
+import io.temporal.workflow.Promise;
+import io.temporal.workflow.Workflow;
+
+import org.apache.gobblin.temporal.util.nesting.work.WFAddr;
+import org.apache.gobblin.temporal.util.nesting.work.Workload;
+
+
+/** Core skeleton of {@link NestingExecWorkflow}: realizing classes need only define {@link #launchAsyncActivity} */
```
Comment (Will-Lo) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/workflow/AbstractNestingExecWorkflowImpl.java:40]: Can we have a bit more documentation on what values are expected for maxBranchesPerTree, maxSubTreesPerTree, etc for the expectation of a more balanced tree?

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/workflow/NestingExecWorkflow.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ * IMPORTANT: `Math.sqrt(maxBranchesPerTree) == maxSubTreesPerTree` provides a good rule-of-thumb; `maxSubTreesPerTree
+ * must not exceed that.  This enables consolidation, wherein continued expansion occurs only along the tree's right-most edges.
+ *
+ * @param <WORK_ITEM> the type of task for which to invoke an appropriate activity
+ * @param maxSubTreesForCurrentTreeOverride when the current tree should use different max sub-trees than descendants
+ */
+@WorkflowInterface
+public interface NestingExecWorkflow<WORK_ITEM> {
+  @WorkflowMethod
+  int performWorkload(
```
Comment (Will-Lo) [inline gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/workflow/NestingExecWorkflow.java:46]: Can we add a javadoc for what the return value should represent?

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/SeqBackedWorkSpan.java (hunk 1 above)
Comment (phet) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/SeqBackedWorkSpan.java:56]: That's the `Iterator` contract I've always known, essentially just a modernization on bounds checking from C, et. al. without that discipline, how could one safely work with an empty `Iterator`?

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/WFAddr.java (hunk 2 above)
Comment (phet) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/work/WFAddr.java:34]: fair enough... good call actually!

File: gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/workflow/AbstractNestingExecWorkflowImpl.java (hunk 3 above)
Comment (phet) [reply gobblin-temporal/src/main/java/org/apache/gobblin/temporal/util/nesting/workflow/AbstractNestingExecWorkflowImpl.java:40]: I situated that advice in the `NestingExecWorkflow` interface - https://github.com/apache/gobblin/pull/3811/files#diff-3ca08118c269ccea822f3ebe53b22f513b848369cae1c8ca52f18fc22f200b59R37 (see: line 37)

**Review by Will-Lo — APPROVED** [review]

**Review by Will-Lo — APPROVED** [review]

**Pull request merged**